### ISSUE TITLE
"Overloaded function signatures overlap with incompatible return types" false positive when there can be a subtypeof both return types

### ISSUE THREAD

**Issue by DetachHead** [python/mypy#13611]
```py
from typing import overload, Literal

class A: ...

class B: ...

class C(A, B): ...

@overload
def foo(a: Literal[True], b: Literal[True]) -> C: ...

@overload
def foo(a: Literal[True], b: bool) -> A: ... # error: Overloaded function signatures 2 and 3 overlap with incompatible return types  [misc]

@overload
def foo(a: bool, b: Literal[True]) -> B: ...


def foo(a: bool, b: bool) -> object: ...
```
[playground](https://mypy-play.net/?mypy=latest&python=3.10&flags=show-error-context%2Cshow-error-codes%2Cstrict%2Ccheck-untyped-defs%2Cdisallow-any-decorated%2Cdisallow-any-expr%2Cdisallow-any-explicit%2Cdisallow-any-generics%2Cdisallow-any-unimported%2Cdisallow-incomplete-defs%2Cdisallow-subclassing-any%2Cdisallow-untyped-calls%2Cdisallow-untyped-decorators%2Cdisallow-untyped-defs%2Cno-implicit-optional%2Cno-implicit-reexport%2Clocal-partial-types%2Cstrict-equality%2Cwarn-incomplete-stub%2Cwarn-redundant-casts%2Cwarn-return-any%2Cwarn-unreachable%2Cwarn-unused-configs%2Cwarn-unused-ignores&gist=1dfc789ebff5ef7ef519d69bebd85867)
